File: 2.1.H/Python/distanceBased.py
import math
from matchStrategy import MatchStrategy

class DistanceBase(MatchStrategy):
    def match(self, individuals, target):

        # drop target itself in individuals first

        for individual in individuals:
            if individual.getId() == target.getId():
                individuals.remove(individual)
                break

        # sort individuals by distance
        individuals.sort(key=lambda individual: self.distance(individual.getCoordinate(), target.getCoordinate()))
        return individuals

    def distance(self, individual1, individual2):
        distance = math.sqrt((individual1.x - individual2.x) ** 2 + (individual1.y - individual2.y) ** 2)
        return distance

File: 2.1.H/Java/StrategyPattern/src/DistanceBased.java
import java.util.List;

public class DistanceBased implements MatchStrategy{
    @Override
    public List<Individual> match(List<Individual> individuals, Individual targetIndividual) {
        //drop target itself in individuals first
        individuals.remove(targetIndividual);

        // Sort individuals by the distance between targetIndividual and each individual
        individuals.sort((individual1, individual2) -> {
            double distance1 = Math.sqrt(Math.pow(individual1.getCoordinate().getX() - targetIndividual.getCoordinate().getX(), 2) + Math.pow(individual1.getCoordinate().getY() - targetIndividual.getCoordinate().getY(), 2));
            double distance2 = Math.sqrt(Math.pow(individual2.getCoordinate().getX() - targetIndividual.getCoordinate().getX(), 2) + Math.pow(individual2.getCoordinate().getY() - targetIndividual.getCoordinate().getY(), 2));
            return (int) (distance1 - distance2);
        });

        individuals.add(0, targetIndividual);
        return individuals;
    }
}
